### PR TITLE
Travis update: 🍏+🐧 use latest shellcheck binary

### LINKED ISSUE #107: Latest shellcheck complaints about main.sh
Folks,

It looks like the latest version of shellcheck complains about a few things in main.sh.  Maybe SC#### numbers have changed for some of these things that might need to be skipped?

`
$ shellcheck main.sh 

In main.sh line 65:
  local color_debug="\x1b[35m"
                     ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 67:
  local color_info="\x1b[32m"
                    ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 69:
  local color_notice="\x1b[34m"
                      ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 71:
  local color_warning="\x1b[33m"
                       ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 73:
  local color_error="\x1b[31m"
                     ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 75:
  local color_critical="\x1b[1;31m"
                        ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 77:
  local color_alert="\x1b[1;33;41m"
                     ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 79:
  local color_emergency="\x1b[1;4;5;33;41m"
                         ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 84:
  local color_reset="\x1b[0m"
                     ^-- SC1117: Backslash is literal in "\x". Prefer explicit escaping: "\\x".

In main.sh line 86:
  if [[ "${NO_COLOR:-}" = "true" ]] || ( [[ "${TERM:-}" != "xterm"* ]] && [[ "${TERM:-}" != "screen"* ]] ) || [[ ! -t 2 ]]; then
                                       ^-- SC2235: Use { ..; } instead of (..) to avoid subshell overhead.

In main.sh line 252:
        ((OPTIND+=__b3bp_tmp_has_arg_${__b3bp_tmp_opt}))
        ^-- SC1105: Shells disambiguate (( differently or not at all. For subshell, add spaces around ( . For ((, fix parsing errors.
          ^-- SC2030: Modification of OPTIND is local (to subshell caused by (..) group).

In main.sh line 269:
  shift $((OPTIND-1))
           ^-- SC2031: OPTIND was modified in a subshell. That change might be lost.

In main.sh line 383:
info "$(echo -e "multiple lines example - line #1\nmultiple lines example - line #2\nimagine logging the output of 'ls -al /path/'")"
                                                 ^-- SC1117: Backslash is literal in "\n". Prefer explicit escaping: "\\n".
                                                                                   ^-- SC1117: Backslash is literal in "\n". Prefer explicit escaping: "\\n".
`

### PR DESCRIPTION
**Pull request by virgilwashere**
# TL;DR

Travis CI tests are successfully passing for both 🐧 Linux & 🍏 OSX.

- [x] assertions
- [x] shellcheck
- [x] `style.pl`

## Changes

- [x] 💚 use latest shellcheck binary 🍏+🐧
  - no more `cabal` installs

- [x] 🔨 use `addons: homebrew` for 🍏 `darwin` (GNU) packages

    ```yaml
    addons:
    apt:
        update: true
    homebrew:
        update: true
        packages:
        - coreutils
        - gnu-sed
        - gnu-tar
        - grep
        - make
        - gawk
        - findutils
        - gnu-which
        - gnu-time
    ```

- [x] 🔨 🍏 add gnubin dirs to $PATH

  - `homebrew` decided to [remove `--with-default-names`][gnu-homebrew]

    ```bash
    if [ "$TRAVIS_OS_NAME" = "osx" ]; then
      if type brew &>/dev/null; then
        HOMEBREW_PREFIX=$(brew --prefix)  # add gnubin commands to PATH
        for d in ${HOMEBREW_PREFIX}/opt/*/libexec/gnubin; do export PATH=$d:$PATH; done
      fi
    fi
    ```

- [x] 👷 limit git clone depth (3)
- [x] 🔨 move `blah --version` debug commands to `before_script`
- [x] 💚 remove `cache: apt` because docs
  - https://docs.travis-ci.com/user/caching/#things-not-to-cache
- [ ] ⭕️ ~~set `dist: trusty` in matrix~~

[gnu-homebrew]: <https://discourse.brew.sh/t/why-was-with-default-names-removed/4405>

### CODE REVIEW

**Comment by kvz** [comment]
Is it correct that this is still a draft?

**Comment by kvz** [comment]
Wondering if this is ready for review?

**Comment by virgilwashere** [comment]
I'm not sure what state the travis tests are in...

**Comment by kvz** [comment]
Probably merge master into your branch and push, that brings a lot of test fixes in, and also kicks a new travis build

**Comment by virgilwashere** [comment]
```bash
      if type brew &>/dev/null; then
        HOMEBREW_PREFIX=$(brew --prefix)
        # gnubin;
        for d in ${HOMEBREW_PREFIX}/opt/*/libexec/gnubin; do export PATH=$d:$PATH; done
        # gnuman
        # for d in ${HOMEBREW_PREFIX}/opt/*/libexec/gnuman; do export MANPATH=$d:$MANPATH; done
      fi
```

**Comment by kvz** [comment]
I'm not entirely clear what this PR brings. Faster installs of shellcheck on both platforms? Although i think linting on just linux suffices, I di like faster installs :)

It also makes git faster by limiting the depth and adopts more travis best practices, but i'm a bit puzzled on the gnu prefix thing still, also commented in #123.  

As far as i know, b3bp does not rely on gnu-specific bins, and recommends to use interoperably supported arguments, _or_ switches between gnu/bsd style, which also happens in acceptance.sh for instance. Likely I'm overlooking something!

**Comment by virgilwashere** [comment]
You can't install gnu-sed as sed easily anymore.  Adding the PATH for each of the homebrew targets is all you can do easily.... unless you want to be checking for gsed, ginstall, ggrep, gsha256sum etc.

Originally,  this PR was addressing #107. Now it does as you say: best practices,  faster CI startup, and adds shellcheck to osx test runs.

In `test/acceptance.sh` there is logic to check for OSX GNU binaries.

**Comment by kvz** [comment]
I see yes. I actually don't think 'overwriting' mac's utils with GNU is good as mac scripts could be relying on bsd versions being available globally. And also since b3bp doesn't rely on such magic we can safely remove that from the PR i think. 

The other changes i'm on board with!

**Comment by kvz** [comment]
From: https://github.com/kvz/bash3boilerplate/issues/123

> I needed to account for the OSX G binaries in the travis scripts... or find a way to make homebrew gnu utils the same as Ubuntu. End result is 1 osx if statement for PATH and the rest of the scripts work on both OS

So you're saying we do need the gnu versions available 'globally' in order for travis scripts to function? Which ones exactly?

**Comment by kvz** [comment]
Closing this due to inactivity. Feel free to re-open!
